Flag trigger + audit logs (#39)
bucketeer-io/bucketeer-docs: docs/feature-flags/creating-feature-flags/trigger.mdx

Author: @cre8ivejp, when I access the Bucketeer portal, I don't have access to the Triggers tab.

![image](https://github.com/bucketeer-io/bucketeer-docs/assets/85108199/21fb3a72-3572-48f6-9d59-ddb5e36b38b5)

---
title: Triggers
# sidebar_position:
slug: /feature-flags/creating-feature-flags/triggers
description: Use the flag triggers to automatically turn feature flags on or off by calling a URL using the POST method.
tags: ['guide', 'automation', 'trigger', 'feature-flag']
---

import CenteredImg from '@site/src/components/centered-img/CenteredImg';

Flag triggers are a feature that Bucketeer provides to let you change the flag's state (ON/OFF) by calling a URL. As a result, you can use this option to enable or disable flags during CI/CD (Continuous Integration/Continuous Deployment) pipelines or based on an APM (Application Performance Monitoring) integration.

> To access the trigger page on the Bucketeer dashboard, select the **Feature Flags** tab, choose the desired flag, and click on its name. Select the **Trigger** tab at the top of the new page. You can check the existing triggers related to the current feature flag.

Reviewer: I think we can simplify this `To access the trigger page on the Bucketeer dashboard, select the **Feature Flags** tab,` by just saying `On the Feature Flag details page`.

Also, `Select the **Trigger** tab at the top of the new page` is not a new page.

:::note Future releases

Bucketeer has in its roadmap the objective to accept payloads in the Webhook body to enable you to provide additional information regarding the request source. Stay tuned for future releases.

:::

## Manage triggers

When you access the Trigger tab, you will have access to all existing triggers related to the current flag. The image provides an example of a flag associated with three triggers.

<CenteredImg
  imgURL="img/feature-flags/trigger/flag-trigger-existing-triggers.png"
/>

Each trigger presents the following information:

- **Action**: Specify if the trigger turns the flag ON or OFF when activated.
- **Trigger URL**: Part of the Trigger URL. As mentioned before, you only have access to the complete trigger URL at the time of trigger creation.
- **Triggered Times**: Inform the number of times the trigger was used and the associated action was performed. 
- **Last Triggered**: Inform the date and time the trigger was last used.

:::info Triggered execution

> The Triggered Times and Last Triggered will only be updated if the trigger action changes the flag state. For instance, if the flag was initially ON, and the trigger was executed to turn it OFF, then the Triggered Times and Last Triggered will be updated. However, suppose the flag was initially OFF, and the trigger action was also intended to turn it OFF. In that case, there will be no changes to the Triggered Times and Last Triggered because the flag state was not changed.

Reviewer: The behavior has changed. When the flag is already enabled, and it tries to enable it, and vice-versa, it will just update the trigger data, including its count. So I think we can remove this information. Only when the trigger is disabled, if the URL is called, it won't do anything.

::: 

After creating a trigger, you can perform four actions on it through the Bucketeer dashboard by clicking on the ellipsis (**...**) located at the top right corner of the trigger card:

- **Edit**: Use this option to update the trigger description. 
- **Disable**: Deactivate the trigger without removing or changing the Trigger URL. Despite receiving a POST request, the Bucketeer system will not execute the trigger action when disabled.
- **Reset**: Create a new Trigger URL. POST requests using the old Trigger URL will not change the flag state. When you Reset the trigger, the Bucketeer system will provide a new Trigger URL. You should copy and save the Trigger URL to use in your application. 
- **Delete**: Remove the trigger.

:::info Reset operation

Resetting a trigger does not restart Triggered Times and Last Triggered information.

> A new field shows up on the trigger card, informing the date of the last reset.

Author: Replaced by:

`If you reset a trigger, the reset date will be displayed on the trigger card. This way, you can quickly check if the trigger was reset before and when it was done.`

One common integration of triggers is with APM solutions, such as:

- [Datadogs](https://www.datadoghq.com/)
- [New Relic One](https://newrelic.com/)
- [Honeycomb](https://www.honeycomb.io/)
- [Dynatrace](https://www.dynatrace.com/)
- [Splunk Observability Cloud](https://www.splunk.com/en_us/products/observability.html)

APMs can automatically modify the flag state based on certain conditions within that platform. Typically, APMs use triggers to shut down flags in pre-configured scenarios. For example, you can use an APM to monitor your system. If the error rate reaches 5%, the APM will call the trigger URL to disable (modify to OFF state) the feature flag.

CI/CD pipelines can also take advantage of flag triggers. You can use triggers when deploying code changes to turn a flag on or off automatically at the final stage of the pipeline. In addition, you can create your own automation algorithm and use triggers to control the flags in your application.

## Create a triggers

> The creation of triggers is a straightforward process. Follow the steps below:

Reviewer: ```suggestion
Creating a trigger is a straightforward process. Follow the steps below:
```

> 1. Access **Triggers** tab.

Reviewer: ```suggestion
1. Access the **Triggers** tab.
```

- **Last Triggered**: Inform the date and time the trigger was last used.
- **Last update**: Inform you about the last time the trigger was updated. This field updates in case of triggers, reset URL, disable/enable, and description update events.   

:::info Triggered execution

The Triggered Times and Last Triggered will be updated every time the trigger request is received, even if the flag state doesn't change. For instance, Triggered Times and Last Triggered will be updated even in the case when a trigger tries to turn OFF a flag that is already OFF. 

Triggered Times and Last Triggered aren't updated, only in case the trigger is disabled when the trigger request is received.

::: 

After creating a trigger, you can perform four actions on it through the Bucketeer dashboard by clicking on the **Action** menu (**...**) located at the top right corner of the trigger card:

- **Edit**: Use this option to update the trigger description. 
- **Disable**: Deactivate the trigger without removing or changing the Trigger URL. Despite receiving a POST request, the Bucketeer system will not execute the trigger action when disabled.

> - **Reset**: Create a new Trigger URL. POST requests using the old Trigger URL will not change the flag state. When you Reset the trigger, the Bucketeer system will provide a new Trigger URL. You should copy and store the Trigger URL to use in your application. 

Reviewer: ```suggestion
- **Reset**: Create a new Trigger URL. POST requests using the old Trigger URL will not change the flag state. When you reset the trigger, the Bucketeer system will provide a new Trigger URL. You should copy and store the Trigger URL to use in your application. 
```

- **Delete**: Remove the trigger.

:::info Reset operation

Resetting a trigger doesn't reset Triggered Times and Last Triggered information.

> If you reset a trigger, the reset date will be displayed on the trigger card. This way you can quickly check if a trigger was reseted before and when it was done.

Reviewer: ```suggestion
If you reset a trigger, the reset date will be displayed on the trigger card. This way, you can quickly check if a trigger was reset before and when it was done.
```

:::

## Test a trigger

You can make a POST request using the Trigger URL to test a trigger. The code block below presents an example of the code used to perform a request using `curl`:

```curl

> https://dev.bucketeer.jp/webhook/MTQyMDY5OTI0ODc2MQw4yPzGndlkezFEsZE_uPnA

Reviewer: ```suggestion
curl -X POST https://api.bucketeer.io/webhook/MTQyMDY5OTI0ODc2MQw4yPzGndlkezFEsZE_uPnA
```
